hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -2189,48 +2156,37 @@ class Emit[C](
 
     def emitVoid(ir: IR, env: E = env, container: Option[AggContainer] = container, loopEnv: Option[Env[LoopRef]] = loopEnv): Code[Unit] = {
       EmitCodeBuilder.scopedVoid(mb) { cb =>
-        this.emitVoid(cb, ir, mb, region, env, container, loopEnv)
+        this.emitVoid(cb, ir, region, env, container, loopEnv)
       }
     }
 
-    def emitStream(ir: IR, outerRegion: ParentStagedRegion): EmitCode =
-      EmitStream.emit(this, ir, mb, outerRegion, env, container)
+    def emitStream(ir: IR, outerRegion: Value[Region] ): EmitCode =
+      EmitCode.fromI(mb)(cb => EmitStream.produce(this, ir, cb, outerRegion, env, container))
 
     val pt = ir.pType
 
     // ideally, emit would not be called with void values, but initOp args can be void
     // working towards removing this
     if (pt == PVoid)
       return EmitCode.fromI(mb) { cb =>
-        this.emitVoid(cb, ir, mb, region, env, container, loopEnv)
+        this.emitVoid(cb, ir, region, env, container, loopEnv)
         IEmitCode.present(cb, PCode._empty)
       }
 
     val result: EmitCode = (ir: @unchecked) match {
 
-      case Let(name, value, body) => value.pType match {
-        case streamType: PCanonicalStream =>
-
-          val separateRegions = ctx.smm.lookup(value).separateRegions
-          assert(separateRegions == streamType.separateRegions)
-          val outerRegion = region.asParent(separateRegions, "Let value")
-          val valuet = emitStream(value, outerRegion)
-          val bodyEnv = env.bind(name -> new EmitUnrealizableValue(streamType, valuet))
-
-          emit(body, env = bodyEnv)
-
-        case valueType =>
-          val x = mb.newEmitField(name, valueType)
-          EmitCodeBuilder.scopedEmitCode(mb) { cb =>
-            x.store(cb, emit(value))
-            val bodyenv = env.bind(name, x)
-            emit(body, env = bodyenv)
-          }
-      }
+      case Let(name, value, body) =>
+        EmitCode.fromI(mb) { cb =>
+          val xVal = if (value.typ.isInstanceOf[TStream]) emitStream(value, region) else emit(value)
```

Comment:
Am I wrong in thinking emitStream can now be folded into emit?

Reply:
Yeah, it can. I'd like to save this for a smaller PR on top of this if possible, though.